File: src/main/java/com/moira/itda/domain/common/service/CommonCodeService.java
package com.moira.itda.domain.common.service;

import com.moira.itda.domain.common.dto.request.CommonCodeAddRequest;
import com.moira.itda.domain.common.dto.request.CommonCodeDetailAddRequest;
import com.moira.itda.domain.common.dto.response.CommonCodeDetailResponse;
import com.moira.itda.domain.common.dto.response.CommonCodeResponse;
import com.moira.itda.domain.common.mapper.CommonCodeMapper;
import com.moira.itda.global.entity.CommonCode;
import com.moira.itda.global.entity.CommonCodeDetail;
import com.moira.itda.global.exception.ErrorCode;
import com.moira.itda.global.exception.ItdaException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@RequiredArgsConstructor
@Service
public class CommonCodeService {
    private final CommonCodeMapper commonCodeMapper;

    /**
     * 공통코드 > 저장
     */
    @Transactional
    public void add(CommonCodeAddRequest request) {
        // [1] 유효성 검사
        if (commonCodeMapper.selectCommonCodeKeyChk(request.codeKey()) > 0) {
            throw new ItdaException(ErrorCode.ALREADY_EXISTING_COMMON_CODE);
        }

        // [2] CommonCode 저장
        CommonCode commonCode = CommonCode.fromCommonCodeAddRequest(request);
        commonCodeMapper.insertCommonCode(commonCode);

        // [3] CommonCodeDetail 저장
        for (CommonCodeDetailAddRequest detailRequest : request.details()) {
            CommonCodeDetail detail = CommonCodeDetail.fromCommonCodeDetailAddRequest(request.codeKey(), detailRequest);
            commonCodeMapper.insertCommonCodeDetail(detail);
        }
    }

    /**
     * 공통코드 > 단건 조회
     */
    @Transactional(readOnly = true)
    public List<CommonCodeDetailResponse> get(String key) {
        // [1] 유효성 검사
        if (commonCodeMapper.selectCommonCodeKeyChk(key) < 1) {
            throw new ItdaException(ErrorCode.COMMON_CODE_NOT_FOUND);
        }

        // [2] 조회
        return commonCodeMapper.selectCommonCode(key);
    }

    /**
     * 공통코드 > 전체 조회
     */
    @Transactional(readOnly = true)
    public List<CommonCodeResponse> getAll() {
        return commonCodeMapper.selectCommonCodeList();
    }

    /**
     * 공통코드 > 삭제
     */
    @Transactional
    public void delete(String key) {
        // [1] 유효성 검사
        if (commonCodeMapper.selectCommonCodeKeyChk(key) < 1) {
            throw new ItdaException(ErrorCode.COMMON_CODE_NOT_FOUND);
        }

        // [2] CommonCodeDetail -> CommonCode 삭제
        commonCodeMapper.deleteCommonCodeDetail(key);
        commonCodeMapper.deleteCommonCode(key);
    }
}

File: src/main/kotlin/com/moira/itda/domain/common_code/service/CommonCodeService.kt
package com.moira.itda.domain.common_code.service

import com.moira.itda.domain.common_code.dto.response.AllCodeDetailResponse
import com.moira.itda.domain.common_code.dto.response.CodeDetailResponse
import com.moira.itda.domain.common_code.mapper.CommonCodeMapper
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class CommonCodeService(
    private val mapper: CommonCodeMapper
) {
    /**
     * 공통 > 공통코드 조회
     */
    @Transactional(readOnly = true)
    fun getCode(key: String): List<CodeDetailResponse> {
        return mapper.selectCodeDetailList(key = key)
    }

    /**
     * 공통 > 공통코드 일괄 조회
     */
    @Transactional(readOnly = true)
    fun getAllCode(keyList: List<String?>): AllCodeDetailResponse {
        val response = AllCodeDetailResponse(null, null, null, null, null)

        keyList.forEachIndexed { index, key ->
            if (key != null) {
                when (index) {
                    0 -> response.code1 = mapper.selectCodeDetailList(key = key)
                    1 -> response.code2 = mapper.selectCodeDetailList(key = key)
                    2 -> response.code3 = mapper.selectCodeDetailList(key = key)
                    3 -> response.code4 = mapper.selectCodeDetailList(key = key)
                    4 -> response.code5 = mapper.selectCodeDetailList(key = key)
                }
            }
        }

        return response
    }
}